core/src/main/scala/org/apache/spark/storage/BlockManagerMasterHeartbeatEndpoint.scala

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage
+
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rpc.{RpcCallContext, RpcEnv, ThreadSafeRpcEndpoint}
+import org.apache.spark.storage.BlockManagerMessages.{BlockManagerHeartbeat, RegisterBlockManager, RemoveExecutor, UpdateBlockInfo}
+
+/**
+ * Separate heartbeat out of BlockManagerMasterEndpoint due to performance consideration.
+ */
+private[spark] class BlockManagerMasterHeartbeatEndpoint(
+    override val rpcEnv: RpcEnv,
+    isLocal: Boolean)
+  extends ThreadSafeRpcEndpoint with Logging {
+
+  // Mapping from block manager id to the block manager's information.
+  private val blockManagerLastSeen = new mutable.HashMap[BlockManagerId, Long]
+
+  // Mapping from executor ID to block manager ID.
+  private val blockManagerIdByExecutor = new mutable.HashMap[String, BlockManagerId]
+
+  override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case RegisterBlockManager(blockManagerId, _, _, _, _) =>
+      updateLastSeenMs(blockManagerId)
+      blockManagerIdByExecutor(blockManagerId.executorId) = blockManagerId
+      context.reply(true)
+
+    case UpdateBlockInfo(blockManagerId, _, _, _, _) =>
+      updateLastSeenMs(blockManagerId)
+      context.reply(true)
+
+    case RemoveExecutor(execId) =>
+      blockManagerIdByExecutor.get(execId).foreach(blockManagerLastSeen.remove)
+      blockManagerIdByExecutor -= execId
+      context.reply(true)
```

Comment:
It would be better if you could comment that why you needs to handle `RegisterBlockManager`, `UpdateBlockInfo`, `RemoveExecutor` besides  `BlockManagerHeartbeat` in a heartbeat related endpoint.

Reply:
thanks, I will add some comments

Reply:
https://github.com/apache/spark/blob/cb6c4f6842ae977c4f664703c1f3df5b23a0a004/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterHeartbeatEndpoint.scala#L40